session/general_log.go

```diff
@@ -0,0 +1,196 @@
+// The entry will be dropped once the channel is full
+func putGeneralLogOrDrop(entry *generalLogEntry) {
+	select {
+	case globalGeneralLogger.logEntryChan <- entry:
+	default:
+		// When logEntryChan is full, the system resource is under so much pressure that the logging system capacity
+		// is reduced. We should NOT output another warning log saying that we are dropping a general log, which will
+		// add more pressure on the system load.
+		generalLogDroppedEntry.Inc()
+		glEntryPool.Put(entry)
+	}
+}
+
+// InitGeneralLog initialize general query logger, which will starts a format & logging worker goroutine
+// General query logs are sent to the worker through a channel, which is asynchronously flushed to logging files.
+func InitGeneralLog() {
+	GeneralLogLogger := newGeneralLogLogger()
+	globalGeneralLogger = newGeneralLogger(GeneralLogLogger)
+}
+
+// StopGeneralLog stops the general log worker goroutine
+func StopGeneralLog() {
+	globalGeneralLogger.stopLogWorker()
+}
+
+// generalLogEntry represents the fields in a general query log
+type generalLogEntry struct {
+	ConnID            uint64
+	User              string
+	SchemaMetaVersion int64
+	TxnStartTS        uint64
+	TxnForUpdateTS    uint64
+	IsReadConsistency bool
+	CurrentDB         string
+	TxnMode           string
+	Query             generalLogEntryQuery
+}
+type generalLogEntryQuery struct {
+	isPrepared      bool
+	originalText    string
+	stmtNode        ast.StmtNode
+	stmtCtx         *stmtctx.StatementContext
+	preparedParams  variable.PreparedParams
+	enableRedactLog bool
+}
+
+// getGeneralLogEntry get a general log entry from the object pool
+func getGeneralLogEntry() *generalLogEntry {
+	return glEntryPool.Get().(*generalLogEntry)
+}
+
+// generalLogger receives general log entry from channel, and log or drop them as needed
+type generalLogger struct {
+	logger       *zap.Logger
+	logEntryChan chan *generalLogEntry
+	quit         chan struct{}
+}
+
+func newGeneralLogger(logger *zap.Logger) *generalLogger {
+	gl := &generalLogger{
+		logger:       logger,
+		logEntryChan: make(chan *generalLogEntry, 10000),
+		quit:         make(chan struct{}),
+	}
+	go gl.startLogWorker()
+	return gl
+}
+
+func (gl *generalLogger) logEntry(e *generalLogEntry) {
+	buf := queryBuilderPool.Get().(*strings.Builder)
+	q := e.Query
+	if q.isPrepared {
+		buf.WriteString(q.originalText)
+	} else {
+		// The following logic should be kept in sync with ExecStmt.GetTextToLog
+		if q.enableRedactLog {
+			sql, _ := q.stmtCtx.SQLDigest()
+			buf.WriteString(sql)
+		} else if sensitiveStmt, ok := q.stmtNode.(ast.SensitiveStmtNode); ok {
+			buf.WriteString(sensitiveStmt.SecureText())
+		} else {
+			buf.WriteString(q.stmtCtx.OriginalSQL)
+			buf.WriteString(q.preparedParams.String())
+		}
+	}
+	bufString := buf.String()
+	queryMutable := *(*[]byte)(unsafe.Pointer(&bufString))
+	for i, b := range queryMutable {
+		if b == '\r' || b == '\n' || b == '\t' {
+			queryMutable[i] = ' '
+		}
+	}
```

Comment:
This is no longer necessary, as the log format will not be broken and we can avoid unsafe.

Reply:
removed.